src/main/java/tr/com/srdc/cda2fhir/transform/ResourceTransformerImpl.java

```diff
@@ -2878,4 +3084,67 @@ public EntryResult tVitalSignObservation2Observation(VitalSignObservation cdaVit
 
 		return result;
 	}
+
+	public Binary tBinary(String encodedBody) {
+		Binary binary = new Binary();
+		binary.setContentElement(new Base64BinaryType(encodedBody));
+		binary.setContentType("text/plain");
+		binary.setId(new IdType("Binary", getUniqueId()));
+		binary.setSecurityContext(new Reference(binary.getId()));
+		return binary;
+	}
+
+	public Device tDevice(Identifier assemblerDevice) {
+		Device device = new Device();
+		device.setStatus(FHIRDeviceStatus.ACTIVE);
+		device.addIdentifier(assemblerDevice);
+
+		Narrative deviceNarrative = new Narrative().setStatus(NarrativeStatus.GENERATED);
+		deviceNarrative.setDivAsString(assemblerDevice.getValue());
+		device.setText(deviceNarrative);
+
+		device.setId(new IdType("Device", getUniqueId()));
+		return device;
+	}
+
+	public Bundle tProvenance(Bundle bundle, String encodedBody, Identifier assemblerDevice) {
```

Comment:
so right now, running files using the default method assembles a bundle that is invalid, because you assemble the bundle even if there is no assembler device. You should add a conditional check to not assemble a bundle if there is no assembler device, as an agent is required per the spec. Basically if there is no assembler device, don't make the bundle (or the binary for that matter).